arch/src/x86_64/mod.rs

```diff
@@ -110,12 +111,18 @@ pub fn configure_system(
 
     let mut params: BootParamsWrapper = BootParamsWrapper(boot_params::default());
 
-    params.0.hdr.type_of_loader = KERNEL_LOADER_OTHER;
-    params.0.hdr.boot_flag = KERNEL_BOOT_FLAG_MAGIC;
-    params.0.hdr.header = KERNEL_HDR_MAGIC;
-    params.0.hdr.cmd_line_ptr = cmdline_addr.raw_value() as u32;
-    params.0.hdr.cmdline_size = cmdline_size as u32;
-    params.0.hdr.kernel_alignment = KERNEL_MIN_ALIGNMENT_BYTES;
+    if setup_hdr.is_some() {
+        params.0.hdr = setup_hdr.unwrap();
+        params.0.hdr.cmd_line_ptr = cmdline_addr.raw_value() as u32;
+        params.0.hdr.cmdline_size = cmdline_size as u32;
+    } else {
+        params.0.hdr.type_of_loader = KERNEL_LOADER_OTHER;
+        params.0.hdr.boot_flag = KERNEL_BOOT_FLAG_MAGIC;
+        params.0.hdr.header = KERNEL_HDR_MAGIC;
```

Comment:
You should be reading this value not setting it. You're missing a major step in the bzImage loading process

I quote from https://www.kernel.org/doc/Documentation/x86/boot.txt

`In 32-bit boot protocol, the first step in loading a Linux kernel
should be to setup the boot parameters (struct boot_params,
traditionally known as "zero page"). The memory for struct boot_params
should be allocated and initialized to all zero. Then the setup header
from offset 0x01f1 of kernel image on should be loaded into struct
boot_params and examined.`

I think the bzImage loader needs to take the zero page in as a parameter and populate it with the necessary fields.)

Reply:
And the reason why should copy them in from the file rather than hardcode them is for future compatibility as new fields may be added with their default value already set

Reply:
@rbradford To have zero page as input to copy setup_header in boot_param only applys to bzimage loader, since elf has no setup_header. So the change should be, pass Some(zero page) for bzimage, None for elf, and remove setup_header from KernelLoaderResult. Agree?

Reply:
Think about it again, actually, compared to the current implementation that return setup_header to  VMM for zero page filling, I indeed don't think that there is any difference with the new change mentioned above. :-P

I will update commit message anyway.

Reply:
@rbradford Thanks for reminding me CONTRIBUTING.md, just refactor commit message following it, since there is no issue opened for this part, not add Fixes # at the end. The following tests are ran and all passed.

cargo build --release
cargo test
cargo fmt --all -- --check
cargo clippy --all -- -D warnings

Reply:
> cargo build --release
> cargo test
> cargo fmt --all -- --check
> cargo clippy --all -- -D warnings

I can see from Travis that the formatting is still wrong. Are you running a new enough version of Rust/Cargo ?

Reply:
> Think about it again, actually, compared to the current implementation that return setup_header to VMM for zero page filling, I indeed don't think that there is any difference with the new change mentioned above. :-P

There is a huge difference. Please make the change that I requested. Some of the fields that you are assigning should be read by the booloader (our bzImage loader) not set. But most importantly the block of data that you copy from the bzImage into the zeropage contains default values that the kernel can look for. Although you might be able to manually set them all for now, new ones can be added in the future.

Further this is the first step of the bzImage loading process and as such I would expect any function that is designed to load such a file would actually carry out this step, rather than delegate it to the VMM.